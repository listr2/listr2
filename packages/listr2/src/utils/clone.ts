import rfdc from 'rfdc'

const clone = rfdc({ circles: true })

/**
 * Deep clones a object in the easiest manner.
 */
// TODO: swap this with structured clone whenever we migrate to node 18
export function cloneObject<T extends Record<PropertyKey, any>>(obj: T): T {
  return clone(obj)
}
