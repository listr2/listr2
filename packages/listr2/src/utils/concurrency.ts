// adapted from https://github.com/fabiospampinato/promise-concurrency-limiter
// with MIT license, since it is again ESM only
// we needed to stop on the first failing case unlike the original
export class Concurrency {
  promise: Promise<void>

  private resolve: () => void
  private reject: (error: Error) => void
  private concurrency: number
  private count: number
  private queue: Set<() => void>

  constructor (options: { concurrency: number }) {
    this.concurrency = options.concurrency
    this.count = 0
    this.queue = new Set()
    this.initPromise()
  }

  public add<T>(fn: () => Promise<T>): Promise<T> {
    if (this.count < this.concurrency) {
      return this.run(fn)
    }

    return new Promise<T>((resolve) => {
      const callback = (): void => resolve(this.run(fn))

      this.queue.add(callback)
    })
  }

  private initPromise (): void {
    const { promise, resolve, reject } = this.withResolvers<void>()

    this.promise = promise
    this.resolve = resolve
    this.reject = reject
  }

  private flush (): void {
    for (const callback of this.queue) {
      if (this.count >= this.concurrency) {
        break
      }

      this.queue.delete(callback)

      callback()
    }
  }

  private run<T>(fn: () => Promise<T>): Promise<T> {
    this.count++

    const promise = fn()

    const cleanup = (): void => {
      this.count--

      this.flush()

      if (this.count === 0) {
        this.resolve()
      }
    }

    promise.then(cleanup, (error) => {
      this.queue.clear()
      this.reject(error)
    })

    return promise
  }

  private withResolvers<T>(): { promise: Promise<T>, resolve: (value: T) => void, reject: (error: Error) => void } {
    let resolve: (value: T) => void
    let reject: (error: Error) => void

    const promise = new Promise<T>((res, rej) => {
      resolve = res
      reject = rej
    })

    return {
      promise,
      resolve: resolve!,
      reject: reject!
    }
  }
}
