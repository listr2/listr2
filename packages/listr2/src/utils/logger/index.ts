export * from './logger.constants'
export type * from './logger.interface'
export * from './logger'
