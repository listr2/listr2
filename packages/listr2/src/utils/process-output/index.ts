export type * from './process-output-buffer.interface'
export * from './process-output-buffer'
export * from './process-output-stream'
export * from './process-output'
export type * from './process-output.interface'
export * from './writable'
