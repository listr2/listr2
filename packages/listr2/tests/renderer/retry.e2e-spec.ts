import { Listr } from '@root'
import type { MockProcessOutput, RendererSetup } from '@tests/utils'
import { expectProcessOutputToMatchSnapshot, mockProcessOutput, unmockProcessOutput, RENDERER_SETUP } from '@tests/utils'

describe.each<RendererSetup>(RENDERER_SETUP)('%s renderer: retry', (renderer, rendererOptions) => {
  const output: MockProcessOutput = {} as MockProcessOutput

  process.stdout.isTTY = true

  beforeEach(async() => {
    mockProcessOutput(output)
  })

  afterEach(async() => {
    unmockProcessOutput(output)
    jest.clearAllMocks()
  })

  // R34CKurUXSpq65S8ebD6jpUrwmIiYrKa
  it('should retry the main task if any of the subtasks fail', async() => {
    let err: Error

    try {
      await new Listr(
        [
          {
            title: 'Some type errors',
            task: async(_, task): Promise<void> => {
              task.output = 'test'

              const retry = task.isRetrying()

              if (retry.count > 0) {
                task.output = `I am self aware that I am retrying for the ${retry.count}th time.`
              }

              throw new Error('This type can not be assigned to type with, oh noes')
            },
            retry: 3
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          renderer,
          rendererOptions
        }
      ).run()
    } catch(e: any) {
      err = e
    }

    expect(err).toBeTruthy()
    expectProcessOutputToMatchSnapshot(output, 'R34CKurUXSpq65S8ebD6jpUrwmIiYrKa')
  })

  it('should stop retrying if the task succeeds afterwards', async() => {
    let err: Error

    try {
      await new Listr(
        [
          {
            title: 'Some type errors',
            task: async(_, task): Promise<void> => {
              const retry = task.isRetrying()

              if (retry?.count === 3) {
                task.title = 'Successed at 3th try.'
              } else {
                throw new Error('not enough')
              }
            },
            retry: 3
          }
        ],
        {
          concurrent: false,
          exitOnError: true,
          renderer,
          rendererOptions
        }
      ).run()
    } catch(e: any) {
      err = e
    }

    expect(err).toBeFalsy()
    expectProcessOutputToMatchSnapshot(output, 'oZPNYsORzcG758b1BLdR91GcErmLlQsw')
  })
})
