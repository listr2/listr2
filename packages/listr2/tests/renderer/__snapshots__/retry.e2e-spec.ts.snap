// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`default renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-exit 1`] = `[]`;

exports[`default renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-stderr 1`] = `[]`;

exports[`default renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-stdout 1`] = `
[
  "[?25l",
  "‚óº Some type errors
",
  "[2K[1A[2K[G‚ùØ Some type errors
",
  "[2K[1A[2K[G‚ùØ Some type errors
  ‚Ä∫ test
",
  "[2K[1A[2K[1A[2K[G‚ö† Some type errors [RETRY:1]
",
  "[2K[1A[2K[G‚ö† Some type errors [RETRY:1]
  ‚Ä∫ test
",
  "[2K[1A[2K[1A[2K[G‚ö† Some type errors [RETRY:1]
  ‚Ä∫ I am self aware that I am retrying for the 1th time.
",
  "[2K[1A[2K[1A[2K[G‚ö† Some type errors [RETRY:2]
  ‚Ä∫ I am self aware that I am retrying for the 1th time.
",
  "[2K[1A[2K[1A[2K[G‚ö† Some type errors [RETRY:2]
",
  "[2K[1A[2K[G‚ö† Some type errors [RETRY:2]
  ‚Ä∫ test
",
  "[2K[1A[2K[1A[2K[G‚ö† Some type errors [RETRY:2]
  ‚Ä∫ I am self aware that I am retrying for the 2th time.
",
  "[2K[1A[2K[1A[2K[G‚ö† Some type errors [RETRY:3]
  ‚Ä∫ I am self aware that I am retrying for the 2th time.
",
  "[2K[1A[2K[1A[2K[G‚ö† Some type errors [RETRY:3]
",
  "[2K[1A[2K[G‚ö† Some type errors [RETRY:3]
  ‚Ä∫ test
",
  "[2K[1A[2K[1A[2K[G‚ö† Some type errors [RETRY:3]
  ‚Ä∫ I am self aware that I am retrying for the 3th time.
",
  "[2K[1A[2K[1A[2K[G‚úñ This type can not be assigned to type with, oh noes
",
  "[2K[1A[2K[G",
  "‚úñ This type can not be assigned to type with, oh noes
",
  "[?25h",
]
`;

exports[`default renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-exit 1`] = `[]`;

exports[`default renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-stderr 1`] = `[]`;

exports[`default renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-stdout 1`] = `
[
  "[?25l",
  "‚óº Some type errors
",
  "[2K[1A[2K[G‚ùØ Some type errors
",
  "[2K[1A[2K[G‚ö† Some type errors [RETRY:1]
",
  "[2K[1A[2K[G‚ö† Some type errors [RETRY:2]
",
  "[2K[1A[2K[G‚ö† Some type errors [RETRY:3]
",
  "[2K[1A[2K[G‚ö† Successed at 3th try. [RETRY:3]
",
  "[2K[1A[2K[G‚úî Successed at 3th try.
",
  "[2K[1A[2K[G",
  "‚úî Successed at 3th try.
",
  "[?25h",
]
`;

exports[`simple renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-exit 1`] = `[]`;

exports[`simple renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-stderr 1`] = `
[
  "‚ö† Some type errors [RETRY:1]
",
  "‚úñ Some type errors [FAILED: This type can not be assigned to type with, oh noes]
",
  "‚ö† Some type errors [RETRY:2]
",
  "‚úñ Some type errors [FAILED: This type can not be assigned to type with, oh noes]
",
  "‚ö† Some type errors [RETRY:3]
",
  "‚úñ Some type errors [FAILED: This type can not be assigned to type with, oh noes]
",
  "‚úñ Some type errors [FAILED: This type can not be assigned to type with, oh noes]
",
]
`;

exports[`simple renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-stdout 1`] = `
[
  "‚ùØ Some type errors
",
  "‚Ä∫ test
",
  "‚Ä∫ test
",
  "‚Ä∫ I am self aware that I am retrying for the 1th time.
",
  "‚Ä∫ test
",
  "‚Ä∫ I am self aware that I am retrying for the 2th time.
",
  "‚Ä∫ test
",
  "‚Ä∫ I am self aware that I am retrying for the 3th time.
",
]
`;

exports[`simple renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-exit 1`] = `[]`;

exports[`simple renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-stderr 1`] = `
[
  "‚ö† Some type errors [RETRY:1]
",
  "‚úñ Some type errors [FAILED: not enough]
",
  "‚ö† Some type errors [RETRY:2]
",
  "‚úñ Some type errors [FAILED: not enough]
",
  "‚ö† Some type errors [RETRY:3]
",
  "‚úñ Some type errors [FAILED: not enough]
",
]
`;

exports[`simple renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-stdout 1`] = `
[
  "‚ùØ Some type errors
",
  "‚úî Successed at 3th try.
",
]
`;

exports[`test renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-exit 1`] = `[]`;

exports[`test renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-stderr 1`] = `
[
  "{"event":"MESSAGE","data":{"retry":{"count":1,"error":{}}},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"error":"This type can not be assigned to type with, oh noes"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"retry":{"count":2,"error":{}}},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"error":"This type can not be assigned to type with, oh noes"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"retry":{"count":3,"error":{}}},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"error":"This type can not be assigned to type with, oh noes"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"error":"This type can not be assigned to type with, oh noes"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
]
`;

exports[`test renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
  "{"event":"OUTPUT","data":"test","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
  "{"event":"TITLE","data":"Some type errors","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
  "{"event":"STATE","data":"RETRY","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"OUTPUT","data":"test","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"OUTPUT","data":"I am self aware that I am retrying for the 1th time.","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"TITLE","data":"Some type errors","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"STATE","data":"RETRY","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"OUTPUT","data":"test","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"OUTPUT","data":"I am self aware that I am retrying for the 2th time.","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"TITLE","data":"Some type errors","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"STATE","data":"RETRY","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"OUTPUT","data":"test","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"OUTPUT","data":"I am self aware that I am retrying for the 3th time.","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"STATE","data":"FAILED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":true,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
]
`;

exports[`test renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-exit 1`] = `[]`;

exports[`test renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-stderr 1`] = `
[
  "{"event":"MESSAGE","data":{"retry":{"count":1,"error":{}}},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"error":"not enough"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"retry":{"count":2,"error":{}}},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"error":"not enough"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"retry":{"count":3,"error":{}}},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"MESSAGE","data":{"error":"not enough"},"task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
]
`;

exports[`test renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-stdout 1`] = `
[
  "{"event":"STATE","data":"STARTED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
  "{"event":"TITLE","data":"Some type errors","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":true,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
  "{"event":"STATE","data":"RETRY","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"TITLE","data":"Some type errors","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"STATE","data":"RETRY","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"TITLE","data":"Some type errors","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"STATE","data":"RETRY","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Some type errors","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"TITLE","data":"Successed at 3th try.","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":false,"isSkipped":false,"hasFinalized":false,"hasSubtasks":false,"title":"Successed at 3th try.","hasReset":true,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":true,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":true,"path":["Some type errors"]}}
",
  "{"event":"STATE","data":"COMPLETED","task":{"hasRolledBack":false,"isRollingBack":false,"isCompleted":true,"isSkipped":false,"hasFinalized":true,"hasSubtasks":false,"title":"Successed at 3th try.","hasReset":false,"hasTitle":true,"isPrompt":false,"isPaused":false,"isPending":false,"isStarted":false,"hasFailed":false,"isEnabled":true,"isRetrying":false,"path":["Some type errors"]}}
",
]
`;

exports[`verbose renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-exit 1`] = `[]`;

exports[`verbose renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-stderr 1`] = `
[
  "[RETRY] Some type errors [1]
",
  "[FAILED] This type can not be assigned to type with, oh noes
",
  "[RETRY] Some type errors [2]
",
  "[FAILED] This type can not be assigned to type with, oh noes
",
  "[RETRY] Some type errors [3]
",
  "[FAILED] This type can not be assigned to type with, oh noes
",
  "[FAILED] This type can not be assigned to type with, oh noes
",
]
`;

exports[`verbose renderer: retry should retry the main task if any of the subtasks fail: R34CKurUXSpq65S8ebD6jpUrwmIiYrKa-stdout 1`] = `
[
  "[STARTED] Some type errors
",
  "[OUTPUT] test
",
  "[TITLE] Some type errors
",
  "[OUTPUT] test
",
  "[OUTPUT] I am self aware that I am retrying for the 1th time.
",
  "[TITLE] Some type errors
",
  "[OUTPUT] test
",
  "[OUTPUT] I am self aware that I am retrying for the 2th time.
",
  "[TITLE] Some type errors
",
  "[OUTPUT] test
",
  "[OUTPUT] I am self aware that I am retrying for the 3th time.
",
]
`;

exports[`verbose renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-exit 1`] = `[]`;

exports[`verbose renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-stderr 1`] = `
[
  "[RETRY] Some type errors [1]
",
  "[FAILED] not enough
",
  "[RETRY] Some type errors [2]
",
  "[FAILED] not enough
",
  "[RETRY] Some type errors [3]
",
  "[FAILED] not enough
",
]
`;

exports[`verbose renderer: retry should stop retrying if the task succeeds afterwards: oZPNYsORzcG758b1BLdR91GcErmLlQsw-stdout 1`] = `
[
  "[STARTED] Some type errors
",
  "[TITLE] Some type errors
",
  "[TITLE] Some type errors
",
  "[TITLE] Some type errors
",
  "[TITLE] Successed at 3th try.
",
  "[COMPLETED] Successed at 3th try.
",
]
`;
